man/rocpmd.1
.if !\n(.g \{\
.   if !\w|\*(lq| \{\
.       ds lq ``
.       if \w'\(lq' .ds lq "\(lq
.   \}
.   if !\w|\*(rq| \{\
.       ds rq ''
.       if \w'\(rq' .ds rq "\(rq
.   \}
.\}
.
.ie \n[.g] .mso www.tmac
.el \{\
. de MTO
\\$2 \(laemail: \\$1 \(ra\\$3
..
. de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.\}

.TH ROCPMD 1

.SH NAME
ROCPMD \- Power management daemon for Red Oak Canyon PowerHAT power management boards
.SH SYNOPSIS
.B rocpmd 
[\fIOPTIONS\fR]

.SH DESCRIPTION
\fBrocpmd\fR is a daemon that interfaces with a Red Oak Canyon PowerHAT power managment expansion board for the 
Raspberry PI computer. The daemon provides a number of functions including
querying of the battery battery level in percent or raw comparator units, logging of 
battery levels to a separate file (rotated on each system start) and the graceful shutdown of the 
Raspberry PI by either pressing the on/off button or when the daemon detects the battery is about to 
be depleted.

.SH STARTING AND STOPPING THE DAEMON
The \fBrocpmd\fR daemon is usually started automatically at boot-up, but can be started manually by 
issuing the following command:

    sudo /etc/init.d/rocpmd.sh start

Issuing the command:

    sudo /etc/init.d/rocpmd.sh stop

will cause \fBrocpmd\fR to stop execution gracefully.

If the daemon is already running and it is necessary to restart it (for example, to re-read the 
configuration file) issue the command:

    sudo /etc/init.d/rocpmd.sh restart


.SH OPTIONS
.SS Generic Daemon Options

.TP
.BR \fB\-\^\-config-path\fR  " <directory path>"
Path of the directory where the configuration file resides (default: /etc), the directory must 
contain a configuration file named \fBrocpmd.conf\fR.

.TP
.BR \fB\-\^\-no-daemon\fR
Do not allow \fBrocpmd\fR to daemonize but run instead as a regular process.

.TP
\fB\-\^\-verbose\fR
.BR
Verbose output during startup. Useful in combination with \fB\-\^\-no-daemon\fR option

.TP
.BR \fB\-\^\-help\fR
Output a usage message and exit.

.SS Power Level Functions

.TP
.BR \fB\-\^\-battery-level\fR
Output the current battery level in percent. The daemon exits immediately afterwards rather than 
going into daemon mode.

.TP
.BR \fB\-\^\-battery-level-raw\fR
Outputs the current battery level in batterly level comparator increments. The daemon exits 
immediately afterwards rather than going into daemon mode.

.TP
.BR \fB\-\^\-power-off\fR
The daemon sends the OFF signal to the power board causing the power to the Raspberry PI to be cut 
immediately and the board to enter the POWER_OFF state. If for some reason the power does not
turn off, the daemon exits immediately afterwards rather than going into daemon mode. 
(This is not a safe shutdown!)

.SH "THE DAEMON CONFIGURATION FILE"
The \fBrocpmd\fR configuration file is written in the 
.URL http://en.wikipedia.org/wiki/YAML/ "YAML"
, human readable data serialzation format. By default the \fBrocpmd\fR configuration file is named 
\fBrocpmd.conf\fR and resides in the /etc directory.  

The configuration file configures four 
functions.  It defines the hardware wiring between the power board signal names and the GPIO port 
numbers, allows for tuning the frequency and timing of the battery level measurement, 
enables or disables battery level logging, and defines a powermap that maps raw battery comparator 
level values to a percentage of battery left.
 

.SS GPIOS

The GPIO (general input/output signal) section of the configuration file defines the mapping and 
configuration (input/output) of the general purpose input/output (GPIO) pins used by the Red Oak 
power managment board. These values do not need to be modified unless the GPIO connections on the 
power board itself have been modified, for example, to resolve GPIO conflicts with other peripherals,
or to adapt the board to a non HAT compliant Raspberry Pi.

.br

The board has some I/O's that are active low and others that are active high. For active high
I/O's the asserted (TRUE) state is represented as a high (3.3V) voltage in the circuit.
For active low I/O's (distinguished by a _B suffix) the asserted (TRUE) state is represented
as a low (0V) voltage in the circuit.

.br

Some I/O's require a pull-up to function correctly.  If a pin requires a pull-up, the 
pin definition in the device-tree blob that loaded from the EEPROM on the PowerHAT will configure
the pin automatically.  If the board has been adapted to different GPIO pins, the new pin mapping 
should be modified in the rocmpd.config and *also* by dtparam= commands placed in /boot/config.txt.  
Please refer to the readme in the device-tree directory of the distribution.

Each entry in the GPIO section of the config file contains three fields:

.br

\fIbmc-pin\fR   - The Broadcom number of the GPIO pin.
.br
\fIrole\fR      - The role of the pin when communicating with the Red Oak PM board.

.br

.TP
\fBLIST OF GPIO PIN ROLES\fR

\fIOFF\fR \- (GPIO output) Toggling this signal high once is logically equivalent to pressing the 
power switch on the board one time.  It causes the internal state machine to advance one step (see 
OPERATION, below).

\fIREQ_OFF_B\fR \- (GPIO input) The power board will assert this signal (low) when the user pushes 
the on/off switch to indicate to \fBrocpmd\fR that the operating system should be shut down and powered 
off.  Note that if the power board is in POWER_ON state, and an OFF is sent to the board, the board state
moves to the REQ_OFF state and asserts REQ_OFF_B.  This is an open drain signal and requires the 
GPIO to have the pull-up configured.

\fICS_B\fR \- (GPIO output) Asserting this signal (low) enables adjusting the battery level 
comparator circuit. (See battery level measurement below.)

\fIUD_B\fR \- (GPIO output) Each time this signal is toggled high the power board's internal battery 
comparator level is actuated once, either up or down. The circuit has 64 discrete comparator levels 
(although all levels are not useful). If UD_B is high when the CS_B signal is asserted (goes low), on 
the next rising edge of UD_B, the comparator level is incremented once. If UD_B is negated (low) when 
the CS_B signal is asserted (goes low), on the next rising edge of UD_B, the comparator level is 
decremented once. (See battery level measurement below.) 

\fIPGOOD\fR \- (GPIO input) This signal will be set high if the power management board's internal 
battery comparator detects that the battery level is above the currently programmed comparator 
level.  It will be low otherwise. This is an open drain signal and requires the GPIO to have the pull-up 
configured.

\fID1_B\fR \- (GPIO input) If the battery is being charged this signal is asserted (set low), if the 
battery is finished charging it is negated (set high).  This I/O requires a pull-up. 

\fID2_B\fR \- (GPIO input) If the PowerHAT is connected to external power this signal 
is asserted (set low), otherwise it is negated (set high).  This I/O requires a pull-up.

.LP

.SS battery-level-reader

This section of the configuration file controls the tuning of the frequency and timing of the 
battery level measurment routine, and enables/disables battery level logging. This section has
three fields.

\fIbattery-level-interval\fR \- The interval in seconds between battery level checks.

\fIbattery-level-gpio-delay\fR \- The delay in microseconds after each time the logic level of 
GPIO signal is changed during battery level checks. 

\fIbattery-level-log\fR \- If set to 'true' a battery level log is written to 
/var/log/rocpmd-battery-level.log. The log contains the date, time, the raw level, and the percent level.
The log file is rotated on each boot, maintaining one previous boot log.

.SS powermap

The powermap is an array of values that map the battery voltage (in compartor level units) output 
by the PowerHAT's internal battery level measurement circuit, to percentages of battery charge 
remaining. The powermap for a specific application and/or battery load condition can be customized 
by using the daemon function \fB\-\^\-battery-level-raw\fR or by enabling \fBbattery-level-log\fR  
to observe the battery discharge curve in compartor units over time, and then establishing the 
percentage battery remaining map (powermap).

The PowerHAT contains a buck/boost (DC-DC) converter that enables the board to supply 5V for 
battery voltages that vary between 4.2V and 2.7V.  Before modifying the map, or using a battery 
other than the one shipped with the power board, the user should uderstand the total system 
current load at 5V (ie, the system power) and ensure that when the battery is in a low voltage 
state (ie, at 2.7V) the current does not exceed the battery's specifications or discharge rate 
(current = power/(2.7V * .85)). (The .85 factor is to account for the conveter efficiency.) Do 
not use a battery with less than 1500mAh capacity.

.SH "OPERATION"

.TP 

.TP
\fBThe Power Board State Machine\fR
The power controller of the Red Oak Canyon PowerHAT is a simple state machine 
driven by the OFF signal or the power button on the board. 
The state machine has three states \fBPOWER_OFF\fR, \fBPOWER_ON\fR, and \fBREQ_OFF\fR: 
.br
.BR
\fB    POWER_OFF\fR \-\-[button press]\-\-> \fBPOWER_ON\fR \-\-[button press or OFF 
toggled high]\-\->
.BR
\fB    REQ_OFF\fR \-\-[putton press or OFF toggled high]\-\-> \fBPOWER_OFF\fR

.TP
\fBPowering on the Raspberry Pi\fR
.br
With the device powered off, the user presses the momentary on/off button causing the PowerHAT 
state to move to the \fBPOWER_ON\fR state and supply power to the Raspberry PI. The raspi 
then boots by power-on-reset. 

.TP
\fBUser initiated shutdown by button\fR
.br
With the Raspberry Pi powered, when the user presses the momentary on/off button, the PowerHAT
progresses to the \fBREQ_OFF\fR state and asserts the REQ_OFF_B signal (low) to the Raspberry PI. 

REQ_OFF_B (request off) is monitored by the \fBrocpmd\fR daemon.  When the daemon detects that the signal asserted 
(went low), it initiates a graceful shutdown of the Linux operating system. As a final step in the 
shutdown sequence, an instance of the daemon running in command mode asserts a final OFF signal to 
the PowerHAT causing the controller to progress from \fBREQ_OFF\fR to the \fBPOWER_OFF\fR state and cut 
power to the Raspberry PI. 

.TP
\fBAutomatic shutdown due to low battery level\fR
.br
When the \fBrocpmd\fR daemon detects a low battery condition via the battery level test, it 
initiates a safe shutdown of the Linux operating system via the system shutdown command. 
As a final step to the shutdown squence, the \fBrocpmd\fR daemon running in command mode asserts 
the OFF signal twice to advance the controller state machine from \fBPOWER_ON\fR, through the \fBREQ_OFF\fR 
state, to the \fBPOWER_OFF\fR state, and cuts power to the Raspberry PI. 

.TP
\fBUser initiated shutdown using the shutdown command\fR
.br
If the OS is shutdown using the shutdown -h -P now command, or, in Jessie, with sudo poweroff, the system 
will shutdown.   As a final step to the shutdown sequence, the \fBrocpmd\fR daemon running in command mode 
asserts the OFF signal twice to advance the controller state machine from \fBPOWER_ON\fR, through the \fBREQ_OFF\fR 
state, to the \fBPOWER_OFF\fR state and cuts power to the Raspberry PI. 

.TP
\fBEmergency Manual Shutdown\fR
.br
If, after pressing the momentary on/off button, the power to the Raspberry PI is not eventuall cut 
for some reason (for example because it never received the OFF signal from \fBrocpmd\fR) a second 
button press will cut the power to the Raspberry PI immediately.  This is not a safe shutdown.

.TP
\fBHow the daemon performs a battery level test\fR
.br
To check the battery level \fBrocpmd\fR first resets the board comparator by setting UD_B  
high, then setting CS_B signal low, and then toggling UD_B 64 times. 
After 64 UD_B pulses, CS_B is negated (set high) again.  This ensures the battery level comparator is
reset to a known condition (railed). 

To establish the battery level, UD_B is set low, then CS_B is asserted (set low), and  
UD_B is toggled until PGOOD changes from high to low. CS_B is then negated (set high). 
The number of UD_B pulses required for the PGOOD signal to go low (the raw value) is then mapped to a 
percentage battery remaining using the power level map array that \fBrocpmd\fR read from the 
configuration file.

During a battery charge cycle, the voltage across the battery terminals does not represent the charge state.  Therefore,
if the battery level is queried during a charge cycle, the daemon will return \fBCharging\fR instead of the battery level or
raw level.  If \fBbattery-level-logging\fR is enabled, the charge cycle is logged as -1 for the battery and 
raw levels.

.SH SEE ALSO
.SS Regular man pages
roc-device-tree(1)

.SS Project Readme
.URL https://github.com/readoakcanyon/HATPowerBoard/README.md

.SS Project sources
The project source code can be obtained at: 
.URL https://github.com/readoakcanyon/HATPowerBoard/

.SS Websites
The Red Oak Canyon web site:
.URL http://redoakcanyon.com



src/rocpmd.cpp
/*
   -----------------------------------------------------------------------------

   Copyright (C) 2015  Kristjan Runarsson, Terry Garyet, Red oak Canyon LLC.

   This file is part of the Red Oak Canyon Power Management Daemon (rocpmd).

   rocpmd is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   rocpmd is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with rocpmd.  If not, see <http:  www.gnu.org/licenses/>.

   -----------------------------------------------------------------------------

   File name: rocpmd.cpp
   Date:      2015-08-25 12:42
   Author:    Kristjan Runarsson

   -----------------------------------------------------------------------------
 */

#include <stdlib.h>
#include <iostream>
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <unistd.h>
#include <signal.h>
#include <time.h>
#include <pthread.h> // Strange things happen to POSIX threads if this is missing.
#include <errno.h>
#include <linux/reboot.h>
#include <sys/reboot.h>
#include <sys/time.h>

//#include "tdaemon.h"
#include "rocpmd.h"
#include "command-line.h"
#include "config.h"
#include "verbose.h"
#include "gpio.h"
//#include "ud_server.h"
#include "ud_client.h"
//#include "logger.h"

#define ROCPMD_LOCKFILE_MISSING       1
#define ROCPMD_LOCKFILE_NOT_READABLE  2
#define ROCPMD_PROCESS_LIVES          3
#define ROCPMD_PROCESS_DEAD           4


using namespace std;

// -----------------------------
// Globals
// -----------------------------

config *g_conf = NULL;
rocpmd *ROCPMD = NULL;

// -----------------------------
// Functions
// -----------------------------

void sig_handler(int signo)
{
    if (signo == SIGINT || signo == SIGHUP || signo == SIGTERM)
    {
        if(g_conf)
        {
            gpio_cleanup(g_conf);
            delete(g_conf);
        }

        if(ROCPMD)
        {
            delete(ROCPMD);
        }

        exit(EXIT_SUCCESS);
    }
}

int rocpmd_instance_lives(string lockfile_path)
{
    // If the lock file does not exist we assume there is no living rocpmd instance.
    if(access(lockfile_path.c_str(), F_OK) < 0)
    {
        return ROCPMD_LOCKFILE_MISSING;
    }

    // If the lock file exists and is not readable we have a problem.
    if(access(lockfile_path.c_str(), R_OK) < 0)
    {
        return ROCPMD_LOCKFILE_NOT_READABLE;
    }

    // ... else we check if the pid in the lockfile is for a living process.

    string pid = "";

    ifstream lockfile(lockfile_path.c_str());

    getline(lockfile, pid);

    lockfile.close();

    int kill_stat = kill(atoi(pid.c_str()), 0);

    /* pid file exists and process lives */ 
    //if(kill(atoi(pid.c_str()), 0) == 0)
    if(kill_stat == 0)
    {
        return ROCPMD_PROCESS_LIVES;
    }

    return ROCPMD_PROCESS_DEAD;
}

void log_and_report(int priority, string prefix, string postfix)
{
	    stringstream msg;
        msg << prefix << postfix;
        cerr << msg.str() << endl;
		syslog(priority, msg.str().c_str());
}

int map_battery_level(config * conf, int battery_level_raw)
{
    int battery_level_percent = -1;

    // The battery level reading is not reliable when the wall wart is connected
    // in which case the gpio_read_battery_level_raw() function will return -1
    // so in the interest of uniformity the percentage power level is also set to
    // -1 to indicate charging status.
    if(battery_level_raw >= 0)
    {
        battery_level_percent = conf->get_powermap_element_at(battery_level_raw);
    }

    return battery_level_percent;
}

bool is_board_installed()
{
    string vendor_path = "/proc/device-tree/hat/vendor";
    string product_id_path = "/proc/device-tree/hat/product_id";

    string error = "";

    error = vendor_path;

    string no_board = " This could be due to a file system error but it most probably means "
                       "that no Red Oak Canyon Power Management Hat board is installed";

    if(access(vendor_path.c_str(), F_OK) == -1)
    {
        error.append(" does not exist.");
        error.append(no_board);
        log_and_report(LOG_CRIT, "Error in board detection: ", error.c_str());
        return false;
    }
    else if(access(vendor_path.c_str(), R_OK) == -1)
    {
        error.append(" is not readable.");
        error.append(no_board);
        log_and_report(LOG_CRIT, "Error in board detection: ", error.c_str());
        return false;
    }

    error = product_id_path;

    if(access(product_id_path.c_str(), F_OK) == -1)
    {
        error.append(" does not exist.");
        error.append(no_board);
        log_and_report(LOG_CRIT, "Error in board detection: ", error.c_str());
        return false;
    }
    else if(access(product_id_path.c_str(), R_OK) == -1)
    {
        error.append(" is not readable.");
        error.append(no_board);
        log_and_report(LOG_CRIT, "Error in board detection: ", error.c_str());
        error.append(no_board);
        return false;
    }

    string wrong_board = " You appear to have the wrong board installed.";

    ifstream vendor_file (vendor_path, fstream::in);
    string vendor = "";
    getline(vendor_file, vendor);
    vendor_file.close();
    vendor = vendor.c_str(); // Gets rid of the \0 character at the end.

    string vendor_expected = "Red Oak Canyon LLC";

    error = "the vendor name of the installed board '";
    error.append(vendor);
    error.append("' is incorrect.");
    error.append(wrong_board);

    int length = min(vendor.length(), vendor_expected.length());

    if(vendor.compare(0, length, vendor_expected) != 0)
    {
        log_and_report(LOG_CRIT, "Board validation failure: ", error);
        return false;
    }

    ifstream product_id_file (product_id_path, fstream::in);
    string product_id = "";
    getline(product_id_file, product_id);
    product_id_file.close();
    product_id = product_id.c_str(); // Gets rid of the \0 character at the end.

    string product_id_expected = "0x0001";

    error = "the product ID of the installed board '";
    error.append(product_id);
    error.append("' is incorrect.");
    error.append(wrong_board);

    length = min(product_id.length(), product_id_expected.length());

    if(product_id.compare(0, length, product_id_expected) != 0)
    {
        log_and_report(LOG_CRIT, "Board validation failure: ", error.c_str());
        return false;
    }

    return true;
}

// -----------------------------
// Classes and structures
// -----------------------------

/*
class rocpmd: public tdaemon
{
    
    public:
    	rocpmd(string daemon_name, string lock_file_name, int daemon_flags);
        ~rocpmd();
    
    protected:
        bool power_log;
        ud_server *ud;
        logger *battery_level_log;
        struct timeval last_read;
    	int daemon_main(config *conf);
        bool init_battery_level_timer();
        int battery_level_raw, battery_level_percent;
        void halt_system();
        bool battery_level_read_interval_expired(config *conf);
};
*/

rocpmd::rocpmd(string daemon_name, string lock_file_name, int daemon_flags)
: tdaemon(daemon_name, lock_file_name, daemon_flags)
{
    gettimeofday(&last_read , NULL);
}

rocpmd::~rocpmd()
{
    delete battery_level_log;
}

int rocpmd::daemon_main(config *conf)
{
    string wd = "";
	check_work_dir(wd);
	
    ud = new ud_server();

    bool create_power_log = conf->get_battery_level_reader().battery_level_log;

    if(create_power_log)
    {
        try
        {
            battery_level_log = new logger("/var/log", "rocpmd-battery-level.log");
        }
        catch(runtime_error re)
        {
            stringstream msg;
    	    msg << "Skipping power level log creation: " << re.what();
            log_and_report(LOG_CRIT, "Exiting: ", msg.str());
            create_power_log = false;
        }
    }

	stringstream msg;
	msg<<"Starting rocpmd (pid: " << get_pid() << ")";
    log_and_report(LOG_CRIT, msg.str(), "");

    battery_level_raw = gpio_read_battery_level_raw(conf);
    battery_level_percent = map_battery_level(conf, battery_level_raw);

    ud->set_battery_level_percent(battery_level_percent);
    ud->set_battery_level_raw(battery_level_raw);

	while(true)
    {
		usleep(100000);

        if(gpio_read_req_off(conf))
        {
            syslog(LOG_CRIT, "Halting due to gpio_read_req_off().");
            halt_system();
        }

        if(battery_level_read_interval_expired(conf))
        {
            battery_level_raw = gpio_read_battery_level_raw(conf);
            battery_level_percent = map_battery_level(conf, battery_level_raw);

            ud->set_battery_level_percent(battery_level_percent);
            ud->set_battery_level_raw(battery_level_raw);

            if(create_power_log)
            {
                try
                {
                    stringstream msg;
                    msg << battery_level_raw << "," << battery_level_percent << "%";
                    battery_level_log->log(msg.str());
                }
                catch(runtime_error re)
                {
                    log_and_report(LOG_CRIT, re.what(), "");
                }
            }
    
            if(battery_level_percent == 0)
            {
                syslog(LOG_CRIT, "Halting due to insufficient battery level.");
                halt_system();
            }
        }
    
        try
        {
            ud->handle_requests();
            if(ud->is_power_off())
            {
                gpio_write_off(conf);
            }
        }
        catch(runtime_error re)
        {
            log_and_report(LOG_CRIT, re.what(), "");
        }
	}

	return 0;
}

void rocpmd::halt_system()
{
    system("sudo /sbin/shutdown -h now");

    if(ROCPMD)
    {
        delete(ROCPMD);
    }

    exit(EXIT_SUCCESS);
}

#define SEC_TO_MSEC(x) x*1000.0

bool rocpmd::battery_level_read_interval_expired(config *conf)
{
    struct timeval now;

    double interval = SEC_TO_MSEC(conf->get_battery_level_reader().battery_level_read_interval);

    double diff, seconds, useconds;

    gettimeofday(&now , NULL);

    seconds  = now.tv_sec  - last_read.tv_sec;
    useconds = now.tv_usec - last_read.tv_usec;

    diff = ((seconds) * 1000 + useconds/1000.0);

    if(diff < interval)
    {
        return false;
    }

    last_read = now;

    return true;
}

// -----------------------------
// Main
// -----------------------------

int main(int argc, char **argv)
{
    bool opt_exit = false;

    signal(SIGINT, sig_handler);
    signal(SIGHUP, sig_handler);
    signal(SIGTERM, sig_handler);

    string daemon_name = "rocpmd";
    string daemon_lockfile_name = daemon_name;
    daemon_lockfile_name.append(".pid");

    cmdline_options opts(argc, argv);

    if(!is_board_installed())
    {
        exit(EXIT_FAILURE);
    }

    if(opts.is_help())
    {
        cmdline_options::usage();
        exit(EXIT_SUCCESS);
    }

    try
    {
		ROCPMD = new rocpmd(daemon_name, daemon_lockfile_name, SINGLETON);
	}
	catch(tdaemon_exception e)
    {
        log_and_report(LOG_CRIT, "Exiting: ", e.what());
        exit(EXIT_FAILURE);
	}
	catch(...)
    {
		unexpected();
	}

    if(geteuid() != 0)
    {
	    stringstream msg;
        msg << "You must have root privileges to run the '" << daemon_name << "' daemon";
        log_and_report(LOG_CRIT, "Exiting: ", msg.str());
        exit (EXIT_FAILURE);
    }

    string config_path;

    if(opts.get_config_path().length())
    {
        config_path = opts.get_config_path();
    }
    
    try
    {        
        g_conf = new config(config_path);
    }
    catch(runtime_error e)
    {
        log_and_report(LOG_CRIT, "Exiting: ", e.what());
        return EXIT_FAILURE;
    }
    catch(...)
    {
        unexpected();
    }

    if(opts.is_verbose())
    {
        verbose_print_options(opts);
        verbose_print_config(g_conf);
    }
    int rocpmd_status = rocpmd_instance_lives(ROCPMD->get_lockfile_path());

    // if( rocpmd_status == ROCPMD_LOCKFILE_MISSING || 
    //     rocpmd_status != ROCPMD_PROCESS_LIVES )
    if(rocpmd_status != ROCPMD_PROCESS_LIVES )
    {
        log_and_report(LOG_CRIT, "Initializing GPIO...","" );
        gpio_init(g_conf);
    }

    if(opts.is_power_off() || 
       opts.is_battery_level() ||
       opts.is_battery_level_raw())
    {
        opt_exit = true;
    }

    if(opts.is_battery_level() || opts.is_battery_level_raw())
    {
        int raw_power = 0;

        rocpmd_status = rocpmd_instance_lives(ROCPMD->get_lockfile_path());

        if(rocpmd_status == ROCPMD_LOCKFILE_NOT_READABLE)
        {
	        stringstream msg;
            msg << "Unable to read the lockfile: " << ROCPMD->get_lockfile_path();
            log_and_report(LOG_CRIT, "Exiting: ", msg.str());
            exit(EXIT_FAILURE);
        }

        //if(rocpmd_status != ROCPMD_LOCKFILE_MISSING || rocpmd_status == ROCPMD_PROCESS_LIVES)
        if(rocpmd_status == ROCPMD_PROCESS_LIVES)
        {
            syslog(LOG_INFO, "Battery read by UD");

            if(opts.is_battery_level_raw())
            {
                int raw_power = ud_client_send_command(ROCPMD_SEND_POWER_LEVEL_RAW);

                if(raw_power >= 0)
                {
                    cout << "Raw power level:        " << raw_power << endl;
                }
                else
                {
                     cout << "Charging..." << endl;
                }
            }

            if(opts.is_battery_level())
            {
                int percent_power = ud_client_send_command(ROCPMD_SEND_POWER_LEVEL_PERCENT);
                
                if(percent_power >= 0)
                {
                    cout << "Percentage power level: " << percent_power << "%" << endl;
                }
                else
                {
                    cout << "Charging..." << endl ;
                }
            }
        }
        else
        {
            syslog(LOG_INFO, "Battery read by SYSFS");

            raw_power = gpio_read_battery_level_raw(g_conf);

            if(raw_power >= 0)
            {
                if(opts.is_battery_level_raw())
                {
                    cout << "Raw power level:        " << raw_power << endl;
                }

                if(opts.is_battery_level())
                {
                    int percent_power = g_conf->get_powermap_element_at(raw_power);
                    cout << "Percentage power level: " << percent_power << "%" << endl;
                }
            }
            else
            {
                cout << "Charging..." << endl;
            }
        }
    }

    if(opts.is_power_off())
    {
        log_and_report(LOG_CRIT, "Powering off ... ", "");
        gpio_write_off(g_conf);
    }

    if(opt_exit && rocpmd_status != ROCPMD_PROCESS_LIVES)
    {
        gpio_cleanup(g_conf);
        exit(EXIT_SUCCESS);
    }

    try
    {
		ROCPMD->run_daemon(g_conf);
	}
	catch(tdaemon_exception e)
    {
        log_and_report(LOG_CRIT, "Exiting: ", e.what());
        exit(EXIT_FAILURE);
	}
	catch(runtime_error e)
    {
        log_and_report(LOG_CRIT, "Exiting: ", e.what());
        exit(EXIT_FAILURE);
	}
	catch(...)
    {
		unexpected();
	}

    gpio_cleanup(g_conf);

    delete g_conf;

	return 0;
} // end main()
